### DISCUSSION TITLE
Should it be possible to call an AWS service within Startup.ConfigureServices?

### DISCUSSION THREAD

**Discussion by lesscodetxm** [aws/aws-lambda-dotnet#1732] [Q&A] [answered]
We found it would be nice to load some app settings from AWS AppConfig at startup, in order to then configure some dependencies in the service collection (eg. logging, Polly retry policies, HttpClientFactory settings, etc). It seems than when we try to access an AWS service from within ConfigureServices, nothing in that method executes after the attempt, and the function is called while missing some dependencies, and obviously fails.

We've confirmed that we can remove the annotations and essentially craft our own "generated" class that makes those calls prior to calling ConfigureServices (and passing the loaded configuration), and it works just fine, so maybe all we need is another hook in Startup (eg `ConfigureEnvironment`) that could allow for this?

**Comment by lesscodetxm** [comment, marked as answer]
Actually, I think part of the failure here was related to a permissions issue when calling AppConfig. After resolving that it seems to work as expected.

**Reply by normj** [reply]
Glad you got it working. You might also try this library for integrating AppConfig into the .NET configuration system. https://github.com/aws/aws-dotnet-extensions-configuration

**Reply by lesscodetxm** [reply]
Yes, this library speaks to the same need - to do work on the _host builder_, before configuring the _services_, analogous to setting up the environment for the function. If you look at their Asp.Net examples, `ConfigureAppConfiguration` comes before `ConfigureServices`. Maybe we could similarly add a new annotation eg `[Configure]` to be applied to another method in Startup?

**Reply by lesscodetxm** [reply]
... or maybe even simpler eg `[AppConfig("AppConfigApplicationId", "AppConfigEnvironmentId", "AppConfigConfigurationProfileId", TimeSpan.FromSeconds(20))]` that grabs the data from AppConfig automatically and makes it available to the function and startup methods as arguments.

Although I'd understand why you'd not want to take a dependency on AppConfig...

**Comment by lesscodetxm** [comment]
Another twist to doing this is that the generated class cannot await the call to ConfigureServices if you define it as `async` in order to await calls to async services.

**Reply by normj** [reply]
I like the idea of making the code generation flexible and supporting the return type of `ConfigureServices` be either `void` or `Task`. Can you submit a feature request on the issues page? Coming from a user helps with prioritization.

**Reply by lesscodetxm** [reply]
I thought about this some more and I'm not sure this isn't an anti-pattern. Since the generated class sets up the service collection from the constructor, it makes it tricky to call an async method there.

I overhauled our use case to use a pure DI approach to loading config using static singleton factory classes that load and cache the config the first time they're used. It's a little bit more verbose but doesn't complicate the Startup logic as much.

Also not sure how this plays into AppConfig's ability to "push" changes to the config for subsequent invocations of the function...
